src/core/metrics.cc

```diff
@@ -368,15 +440,76 @@ void impl::add_registration(const metric_id& id, const metric_type& type, metric
         _value_map[name].info().inherit_type = type.type_name;
         _value_map[name].info().name = id.full_name();
         _value_map[name].info().aggregate_labels = aggregate_labels;
-        _value_map[name][id.labels()] = rm;
+        _value_map[name][rm->info().id.labels()] = rm;
     }
     dirty();
 }
 
+future<bool> impl::set_relabel_configs(std::vector<relabel_config> relabel_configs) {
+    _relabel_configs = std::move(relabel_configs);
+    bool success = true;
+    for (auto&& family : _value_map) {
+        std::vector<shared_ptr<registered_metric>> rms;
+        for (auto&& metric = family.second.begin(); metric != family.second.end();) {
+            metric->second->info().id.labels().clear();
+            metric->second->info().id.labels() = metric->second->info().original_labels;
+            for (auto rl : _relabel_configs) {
+                if (update(rl, metric->second->info())) {
+                    dirty();
+                }
+            }
+            if (metric->first != metric->second->info().id.labels()) {
+                // If a metric labels were changed, we should remove it from the map, and place it back again
+                rms.push_back(metric->second);
+                family.second.erase(metric++);
+                dirty();
+            } else {
+                ++metric;
+            }
+        }
+        for (auto rm : rms) {
+            labels_type lb = rm->info().id.labels();
+            if (family.second.find(rm->info().id.labels()) != family.second.end()) {
+                /*
+                 You can not have a two metrics with the same name and label, there is a problem with the configuration.
+                 On startup we would have throw an exception and stop.
+                 But during normal run, we don't want to crash the server just because a metric reconfiguration.
+                 We cannot throw away the metric.
+                 Instead we add it with an extra label, allowing the user to reconfigure.
+                */
+                seastar_logger.error("Metrics: After relabeling, registering metrics twice for metrics : {}", family.first);
+                success = false;
+                auto id = get_unique_id();
+                lb["err"] = id;
+                rm->info().id.labels()["err"] = id;
+            }
+
+            family.second[lb] = rm;
+        }
+    }
+    return make_ready_future<bool>(success);
```

Comment:
I now understand this bool. Do you undo any changes if you return false?

I think it's better to throw an exception, it's a sort of double registration.

Do you catch the case where you a new metric becomes a double registration after applying labels? I'd guess so, please make sure with a test.

Reply:
on normal situation I would have throw and crash scylla, but this is something that is done in runtime, and I don't wnat to crash the system just becasue someone played with the metrics.
Instead, to prevent double registration we add a label with a unique id.
The user would be able to change the metrics configuration.

It is covered by tests

Reply:
Why would throwing crash scylla? (or better return an exceptional future).

It's just an odd API.

Reply:
@avikivity during metrics registration (usually, at startup) if there are metrics conflict, we throw expecting the process to crash, assuming it's better to crash than to run without all metrics.

The question what to do if someone changes metrics label and cause a metric conflict, we can throw, but at this time, the metrics are already removed from the registration but are not inserted back, so the metric(s) are gone.
During strtup that's ok - because we assume the process would crash.
You cannot roll back that specific metric that has conflict, because there's no promise that would work, we are changing all the metric with the same name at a time and the conflict is found when you insert a metric.

I understand that it's confusing, but I didn't have a better idea, let me know if you want to take it to another direction

Reply:
I think I understand, sometimes you want to throw, sometimes you want to swallow the error. Is this an internal API only?

Reply:
This is an internal API but there will be at least one (if not more) external API that would use it.

Reply:
No, it's never called internally.

How about instead of bool, return a struct.

```c++
struct metric_relabeling_status {
    size_t metrics_ignored_due_to_collision;
};
```

Then the caller, instead of seeing an undocumented `bool`, sees a well explained thing and can take any action they want.

Reply:
@avikivity sure, struct as a documented error is a good idea, I'll do that. That would be on top of the actuall error handling (i.e. add a label with a unique value to the metric so it will be possible to re-register the metric and it will not be lost).

Hence, metrics cannot be ignored or they'll be lost forever, so the name should be something like:
metrics_relabeld_due_to_collision;

Reply:
Sure, that highlights the need for the struct in the first place. There's no way anyone would guess what it means, when the real reason is so subtle.